microceph/api/types/replication.go

```diff
@@ -0,0 +1,125 @@
+package types
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/canonical/lxd/shared/logger"
+	"github.com/canonical/microceph/microceph/constants"
+)
+
+// ################################## Generic Replication Request ##################################
+type ReplicationRequestType string
+
+// This value is split till '-' to get the API request value.
+const (
+	EnableReplicationRequest    ReplicationRequestType = "PUT-" + constants.EventEnableReplication
+	ConfigureReplicationRequest ReplicationRequestType = "PUT-" + constants.EventConfigureReplication
+	DisableReplicationRequest   ReplicationRequestType = "DELETE-" + constants.EventDisableReplication
+	StatusReplicationRequest    ReplicationRequestType = "GET-" + constants.EventStatusReplication
+	ListReplicationRequest      ReplicationRequestType = "GET-" + constants.EventListReplication
+)
+
+type CephWorkloadType string
+
+const (
+	RbdWorkload CephWorkloadType = "rbd"
+	FsWorkload  CephWorkloadType = "cephfs"
+	RgwWorkload CephWorkloadType = "rgw"
+)
+
+type ReplicationRequest interface {
+	GetWorkloadType() CephWorkloadType
+	GetAPIObjectId() string
+	GetAPIRequestType() string
+	GetWorkloadRequestType() string
+}
+
+// Slices
+type MirrorPool struct {
+	Name string
+	Mode RbdResourceType
+}
+
+type MirrorImage struct {
+	Name string
+	Mode RbdReplicationType
+}
+
+type MirrorPools []MirrorPool
+type MirrorImages []MirrorImage
+
+// ################################## RBD Replication Request ##################################
+type RbdResourceType string
+type RbdReplicationDirection string
+
+const (
+	RbdReplicationDirectionRXOnly RbdReplicationDirection = "rx-only"
+	RbdReplicationDirectionRXTX   RbdReplicationDirection = "rx-tx"
+)
+
+const (
+	RbdResourceDisabled RbdResourceType = "disabled"
+	RbdResourcePool     RbdResourceType = "pool"
+	RbdResourceImage    RbdResourceType = "image"
+)
+
+type RbdReplicationType string
+
+const (
+	RbdReplicationDisabled   RbdReplicationType = "disable"
+	RbdReplicationJournaling RbdReplicationType = "journal"
+	RbdReplicationSnapshot   RbdReplicationType = "snapshot"
+)
+
+type RbdReplicationRequest struct {
```

Comment:
As this seems a key type it'd be great to get some docs

Reply:
ack